OpenID Connect dev chapter (#182)
mirumee/saleor-docs: docs/developer/extending/plugins.mdx

Saleor ships with some plugins included, you can find those in the `saleor/plugins/` and `saleor/payment/gateways/` directories of the core.

## Built-in plugins

By default Saleor installs with the following plugins:

- *saleor.plugins.avatax.plugin.AvataxPlugin*
- *saleor.plugins.vatlayer.plugin.VatlayerPlugin*
- *saleor.plugins.webhook.plugin.WebhookPlugin*
- *saleor.plugins.invoicing.plugin.InvoicingPlugin*
- *saleor.payment.gateways.dummy.plugin.DummyGatewayPlugin*
- *saleor.payment.gateways.dummy_credit_card.plugin.DummyCreditCardGatewayPlugin*
- *saleor.payment.gateways.stripe.plugin.StripeGatewayPlugin*
- *saleor.payment.gateways.braintree.plugin.BraintreeGatewayPlugin*
- *saleor.payment.gateways.razorpay.plugin.RazorpayGatewayPlugin*
- *saleor.payment.gateways.adyen.plugin.AdyenGatewayPlugin*

> - *saleor.payment.gateways.authorize_net.plugin.AuthorizeNetGatewayPlugin*

Reviewer: Can we use their official names instead of Python paths here?

## Build-in plugins

By default Saleor lands with these build-in plugins:

- *saleor.plugins.avatax.plugin.AvataxPlugin*
- *saleor.plugins.vatlayer.plugin.VatlayerPlugin*
- *saleor.plugins.webhook.plugin.WebhookPlugin*
- *saleor.payment.gateways.dummy.plugin.DummyGatewayPlugin*
- *saleor.payment.gateways.dummy_credit_card.plugin.DummyCreditCardGatewayPlugin*
- *saleor.payment.gateways.stripe.plugin.StripeGatewayPlugin*
- *saleor.payment.gateways.braintree.plugin.BraintreeGatewayPlugin*
- *saleor.payment.gateways.razorpay.plugin.RazorpayGatewayPlugin*
- *saleor.payment.gateways.adyen.plugin.AdyenGatewayPlugin*
- *saleor.plugins.invoicing.plugin.InvoicingPlugin*
- *saleor.plugins.openid_connect.plugin.OpenIDConnectPlugin*

### OpenIDConnectPlugin
The below charts explain the workflow for actions defined by OpenIDConnectPlugin 

#### External authentication
<Chart
  definition={`
  sequenceDiagram

>     Web ->>+ Saleor: Mutation externalAuthentication with<br> redirect as input (http://localhost:3000/)

Reviewer: ```suggestion
    Web ->>+ Saleor: externalAuthentication mutation with<br> redirect as input (http://localhost:3000/)
```

    Saleor ->>+ OpenIDConnectPlugin: call external_authentication

>     OpenIDConnectPlugin -->>- Saleor: return redirect url to <br>Oauth provider

Reviewer: ```suggestion
    OpenIDConnectPlugin -->>- Saleor: return redirect URL to <br>Oauth provider
```

    Saleor -->>- Web: Response from plugin

>     Web ->>+ OAuthProvider: Redirect to authorizantion page

Reviewer: ```suggestion
    Web ->>+ OAuthProvider: Redirect to the authorizantion page
```

    OAuthProvider -->>- OpenIDConnectPlugin: GET<br> /plugins/<br>mirumee.authentication.openidconnect/<br>callback?state=redirectUrl/&code=YYY
    OpenIDConnectPlugin ->>+ OAuthProvider: fetch tokens
    OAuthProvider -->>- OpenIDConnectPlugin: tokens
    OpenIDConnectPlugin ->> OpenIDConnectPlugin: validate tokens
    OpenIDConnectPlugin ->> OpenIDConnectPlugin: create internal tokens with granted permissions
    OpenIDConnectPlugin -->> Web: set cookie refreshToken with <br>refresh token and redirect to<br> http://localhost:3000/?token=ZZZ&refreshToken=WWW&csrfToken=EEE
`}
/>

#### External refresh
<Chart
  definition={`
  sequenceDiagram

>     Web ->>+ Saleor: mutatuion ExternalRefresh with<br> refreshToken and csrf token as an input*

Reviewer: ```suggestion
    Web ->>+ Saleor: ExternalRefresh mutation with<br> refreshToken and CSRF token as input*
```

    Saleor ->>+ OpenIDConnectPlugin: call <br>PluginManager.external_refresh

>     OpenIDConnectPlugin ->>OpenIDConnectPlugin: validate csrf and <br>refresh token structure

Reviewer: ```suggestion
    OpenIDConnectPlugin ->> OpenIDConnectPlugin: validate CSRF and <br>refresh token structure
```

    OpenIDConnectPlugin ->>+ OAuthProvider: refresh tokens
    OAuthProvider -->>- OpenIDConnectPlugin: new tokens
    OpenIDConnectPlugin ->> OpenIDConnectPlugin: create new internal tokens
    OpenIDConnectPlugin -->>- Saleor: new tokens
    Saleor ->>- Web: return access_token<br> and refresh_token
`}
/>

> \* refresh token can be also passed as a cookie - refreshToken

#### External verify
<Chart
  definition={`
  sequenceDiagram
    Web ->>+ Saleor: mutatuion ExternalVerify with<br> token as an input
    Saleor ->>+ OpenIDConnectPlugin: call <br>PluginManager.external_verify

>     OpenIDConnectPlugin ->>OpenIDConnectPlugin: decode token

Reviewer: ```suggestion
    OpenIDConnectPlugin ->> OpenIDConnectPlugin: decode token
```

    OpenIDConnectPlugin -->>- Saleor: return assigned user<br> and payload data
    Saleor ->>- Web: return token data
`}
/>

#### External logout
<Chart
  definition={`
  sequenceDiagram
    Web ->>+ Saleor: mutatuion ExternalLogout with<br> data to attach as an input
    Saleor ->>+ OpenIDConnectPlugin: call <br>PluginManager.external_logout

>     OpenIDConnectPlugin ->>OpenIDConnectPlugin: prepare logout url:<br> logoutUrl + input data as GET params

Reviewer: ```suggestion
    OpenIDConnectPlugin ->> OpenIDConnectPlugin: prepare logout URL:<br> logoutUrl + input data as GET params
```

    OpenIDConnectPlugin -->>- Saleor: return logoutData
    Saleor ->>- Web: return logoutData
`}
/>

#### Authenticate user
<Chart
  definition={`
  sequenceDiagram

>     Web ->>+ Saleor: Request to protected resources with token

Reviewer: ```suggestion
    Web ->>+ Saleor: request to protected resources with token
```

>     Saleor ->>+ OpenIDConnectPlugin: Call in middleware <br>Plugin.authenticate_user

Reviewer: ```suggestion
    Saleor ->>+ OpenIDConnectPlugin: middleware calls <br>Plugin.authenticate_user
```

>     OpenIDConnectPlugin ->> OpenIDConnectPlugin: Decode and validate token

Reviewer: ```suggestion
    OpenIDConnectPlugin ->> OpenIDConnectPlugin: decode and validate token
```

    OpenIDConnectPlugin ->> OpenIDConnectPlugin: get user
    OpenIDConnectPlugin ->> OpenIDConnectPlugin: assign permissions <br>from token

>     OpenIDConnectPlugin ->>- Saleor: User authenticated

Reviewer: ```suggestion
    OpenIDConnectPlugin ->>- Saleor: user authenticated
```

    Saleor ->>- Web: requested resources
`}
/>

> ### OAuth permissions

Reviewer: Can this be disabled? We should provide a complete list of scopes that need to exist for the plugin to work (the request will likely fail if the auth provider does not know any of the scopes requested).

Author: It is optional. If you don't have defined permissions scopes in your audience, we will get a standard token payload. In that case, each user has empty permissions.
I will edit this section to be more precise.

Reviewer: Are you sure no popular providers will fail the request with "unknown scopes"?

Author: Ech. Thanks for catching this. Again, I've focused on the Auth0.

Saleor can use OAuth permissions assigned to a user. It maps internal permissions to OAuth scopes and attaches 
it to the authorization request.

```python
def _get_oauth_session(self):
    scope = "openid profile email "

>     scope += " ".join([f"saleor:{perm}" for perm in get_permissions_codename()])

Reviewer: I'd rather see a full list of requested scopes than source code here. If we introduce new permissions we will be able to say things like "new since 3.0".

```

> It generates similar requests as below:

Reviewer: ```suggestion
It generates a request similar to the one below:
```

```url
https://saleor.eu.auth0.com/authorize?response_type=code&client_id=...&redirect_uri=..
&scope=openid+profile+email+saleor:manage_users+saleor:manage_staff+saleor:manage_apps
+saleor:manage_apps+saleor:manage_discounts+saleor:manage_plugins
+saleor:manage_gift_card+saleor:manage_menus+saleor:manage_orders+saleor:manage_pages
+saleor:manage_products+saleor:manage_shipping+saleor:manage_settings
+saleor:manage_translations+saleor:manage_checkouts+offline_access&state=...
```

If the user has granted given permission on the OAuth side, Saleor will append the granted 
permission to the authentication token which means that the user will have access to the 
given resources.
```python

> def get_saleor_permissions_from_scope(scope: str) -> List[str]:

Reviewer: Again, no reason to duplicate code in docs.
